ext/drstatecmp/drstatecmp.c

```diff
@@ -0,0 +1,572 @@
+    if (preinsert)
+        instrlist_meta_preinsert(ilist, where, label);
+    else
+        instrlist_meta_postinsert(ilist, where, label);
+    return label;
+}
+
+/* Returns whether or not instr may have side effects. */
+static bool
+drstatecmp_may_have_side_effects_instr(instr_t *instr)
+{
+    /* Instructions with side effects include instructions that write to memory,
+     * interrupts, and syscalls. To avoid inter-procedural analysis, all function calls
+     * are conservatively assumed to have side effects.
+     */
+    return instr_writes_memory(instr) || instr_is_interrupt(instr) ||
+        instr_is_syscall(instr) || instr_is_call(instr);
+}
+
+/****************************************************************************
+ * APPLICATION-TO-APPLICATION PHASE
+ *
+ * In this phase, the side-effect-free basic blocks are duplicated. */
+
+/* Duplicates the bb if it is side-effect-free. */
+static dr_emit_flags_t
+drstatecmp_duplicate_phase(void *drcontext, void *tag, instrlist_t *bb, bool for_trace,
+                           bool translating, OUT void **user_data)
+{
+    /* Allocate space for user_data. */
+    drstatecmp_user_data_t *data = (drstatecmp_user_data_t *)dr_thread_alloc(
+        drcontext, sizeof(drstatecmp_user_data_t));
+    memset(data, 0, sizeof(*data));
+    *user_data = (void *)data;
+
+    /* Determine whether the basic block is free of side effects. */
+    bool side_effect_free = true;
+    for (instr_t *inst = instrlist_first_app(bb); inst != NULL;
+         inst = instr_get_next_app(inst)) {
+        side_effect_free &= !drstatecmp_may_have_side_effects_instr(inst);
+    }
+    /* Cannot easily execute twice the bb if it has side-effects. Thus, no need to
+     * duplicate.
+     */
+    if (!side_effect_free) {
+        data->side_effect_free = false;
+        return DR_EMIT_DEFAULT;
+    }
+    data->side_effect_free = true;
+
+    /* Duplication process.
+     * Consider the following example bb:
+     *   instr1
+     *   instr2
+     *   term_instr
+     *
+     * In this stage, we just duplicate the bb (except for its terminating
+     * instruction) and add special labels to the original and duplicated blocks. The
+     * duplicated bb is for now skipped with a jump. The insert instrumentation stage will
+     * remove this jump, and add saving/restoring of machine state and the state
+     * comparison.
+     * Note that there might be no term_instr (no control transfer instruction) and the bb
+     * just falls-through. Even with no term_instr the jmp and the TERM label are inserted
+     * in the same way, as shown in this example.
+     *
+     * The example bb is transformed, in this stage, as follows:
+     * ORIG_BB:
+     *   instr1
+     *   instr2
+     *   jmp TERM
+     *
+     * COPY_BB:
+     *   instr1
+     *   instr2
+     * TERM:
+     *   term_instr
+     *
+     */
+
+    /* Create a clone of bb. */
+    instrlist_t *copy_bb = instrlist_clone(drcontext, bb);
```

Comment:
If sthg like rep str or scatter expansion went first and inserted emulation labels with stored data, cloning becomes a problem as the clone has the labels but they are missing some data (IIRC the data is set to NULL).  Looks like the late priority will have this go after those expansions.  Better to go first?

Reply:
Hmm, I was thinking it would be good to have the original rather than expanded version (i.e., single scatter instruction; single rep movs insruction), so going first was always good.  But for true emulation where the original won't even run on the underlying hardware: you'd want to go after that.

Reply:
Actually for the rep movs I guess you couldn't take the original which would loop all N iters at once and compare to the expanded version which would do just one iter for each state cmp.

Reply:
Handle app2app transformations with new implementation.